### PR TITLE
Cleanup transaction construction

### PR DESCRIPTION
**Pull request by garious**
#### Problem

Transaction and TransactionBuilder are overreaching. Neither should provide constructors that sign the transaction or set fees or blockhashes. That's the client's job.

#### Summary of Changes

* Working towards transaction modules where constructors don't require keypairs, fee or blockhash parameters.
* Enabled all the signing assertions by default
* Fixed all the bugs caught by those assertions

### CODE REVIEW

File: programs/budget_api/src/budget_transaction.rs [hunk 1]
```diff
@@ -69,11 +72,13 @@ impl BudgetTransaction {
...
     ) -> Transaction {
         let from = from_keypair.pubkey();
-        TransactionBuilder::default()
+        let mut tx = TransactionBuilder::default()
             .push(BudgetInstruction::new_apply_timestamp(
                 &from, contract, to, dt,
             ))
-            .sign(&[from_keypair], recent_blockhash)
+            .compile();
+        tx.sign(&[from_keypair], recent_blockhash);
```
Comment (mvines) [inline programs/budget_api/src/budget_transaction.rs]: wdyt about `.sign()` returning the Transaction?

```rust
    fn foo(...) -> Transaction {
         TransactionBuilder::default()
            .push(BudgetInstruction::new_apply_timestamp(
                &from, contract, to, dt,
            ))
            .compile()
            .sign(&[from_keypair], recent_blockhash)
    }
```

Comment (garious) [reply programs/budget_api/src/budget_transaction.rs]: It won't make sense to do that once we have ThinClient signing pre-existing transactions.

Comment (garious) [reply programs/budget_api/src/budget_transaction.rs]: And more generally, doing a mutable operation and returning the thing you mutated isn't too cool by functional programming standards.  I assume that's why `Vec::sort()` doesn't return itself, for example.

File: sdk/src/transaction_builder.rs [hunk 2]
```diff
@@ -51,6 +50,20 @@ impl TransactionBuilder {
         }
     }
 
+    /// Create a new unsigned transaction from a single instruction
+    pub fn new_singleton(instruction: BuilderInstruction) -> Transaction {
```
Comment (mvines) [inline sdk/src/transaction_builder.rs]: How about we dumb this down as `new_with_instruction()`?  The word "singleton" brings me directly to thinking about the Singleton pattern, which this is not.

Comment (garious) [reply sdk/src/transaction_builder.rs]: done

**Pull request merged**